src/lib/gadgets/range-check.ts

```diff
@@ -48,3 +48,154 @@ function rangeCheck64(x: Field) {
     false // not using compact mode
   );
 }
+
+// default bigint limb size
+const L = 88n;
+const twoL = 2n * L;
+const lMask = (1n << L) - 1n;
+
+/**
+ * Asserts that x, y, z \in [0, 2^88)
+ */
+function multiRangeCheck(x: Field, y: Field, z: Field) {
+  if (x.isConstant() && y.isConstant() && z.isConstant()) {
+    if (x.toBigInt() >> L || y.toBigInt() >> L || z.toBigInt() >> L) {
+      throw Error(`Expected fields to fit in ${L} bits, got ${x}, ${y}, ${z}`);
+    }
+    return;
+  }
+
+  let [x64, x76] = rangeCheck0Helper(x);
+  let [y64, y76] = rangeCheck0Helper(y);
+  rangeCheck1Helper({ x64, x76, y64, y76, z, yz: new Field(0) });
+}
+
+/**
+ * Compact multi-range-check - checks
+ * - xy = x + 2^88*y
+ * - x, y, z \in [0, 2^88)
+ *
+ * Returns the full limbs x, y, z
+ */
+function compactMultiRangeCheck(xy: Field, z: Field): [Field, Field, Field] {
+  // constant case
+  if (xy.isConstant() && z.isConstant()) {
+    if (xy.toBigInt() >> twoL || z.toBigInt() >> L) {
+      throw Error(
+        `Expected fields to fit in ${twoL} and ${L} bits respectively, got ${xy}, ${z}`
+      );
+    }
+    let [x, y] = splitCompactLimb(xy.toBigInt());
+    return [new Field(x), new Field(y), z];
+  }
+
+  let [x, y] = exists(2, () => splitCompactLimb(xy.toBigInt()));
```

Comment:
Just a question for me here :D 
My understanding is that `exists` brings in a witness, just like `Provable.witness` does. When should you use `exists` vs `Provable.witness`?

Reply:
`Provable.witness` handles any provable data types. I created `exists` to have something that's quicker to use if you just want an array of field elements. It's better than `witness` for that in three ways:
* no boilerplate for specifying the type `Provable.Array(Field, 5)`
* you can return bigints from the callback instead of `Fields`
* the array length is typed! i.e. it's a type error to return or unpack too many outputs, like `[x,y,z]`in this example

My thinking was that for low-level gadgets, everything will be arrays of field elements so it makes to have something which optimizes for that

Reply:
Awesome! Thank you for the explanation. I must admit, I used `Provable.witness` for the rotation gadget, not a big deal but I wish I used `exists` instead!